### DISCUSSION TITLE
Memory Issues /  Configuring For Kubernetes

### DISCUSSION THREAD

**Discussion by pinkfloydx33** [dlemstra/Magick.NET#1744] [Help]
# The Problem

We are using Magick.NET as part of an ASP.NET application that runs as a Linux Docker containers in Kubernetes. The application creates thumbnails for any images our users might upload. The maximum upload size is 30Mb. Everything has been working generally fine. However once we get a handful of user uploads (usually in short succession or triggered by QA testing) the pod's memory consumption rises and Kubernetes OOM kills the pod. 

We run our containers with a memory limit of `350Mi`. Because of this, before the OOM's I _originally_ set `ResourceLimits.LimitMemory` to 35% out of precaution. Once we started getting the OOM kills I lowered it even further to 20% which didn't seem to help. I even increased the container max memory to `700Mi` and despite this it eventually gets OOM killed.

I had read some other threads here that suggested making sure you were running as x64 (since we're using the AnyCPU build) and that `ResourceLimits` were configured properly. So I added some basic console logging to check what `ResouceLimits.Memory` was _actually_ defaulting to, and this was the output:

> IMAGEMAGICK MEMORY DEFAULT: 14653595648

Well, `14653595648` bytes equates to more than 13GB which is _well over_ the memory limits set on the pods. It appears that when running in Kubernetes as a Linux container, the defaults are being taken from the `MemTotal` entry in  `/proc/meminfo` (or it's equivalent) which equate to the **node's** resources, and _not_ the pod's. 

Here's the relevant portion of `/proc/meminfo`:

```
MemTotal:       14310152 kB
MemFree:          808760 kB
MemAvailable:    7831240 kB
```

Instead, the defaults should be taken from the cgroups (v2) file at `/sys/fs/cgroup/memory.max`, which in this instance contains

```
367001600
```
correlating to the limit of `350Mi` that this pod currently has.

# Questions

Now that I realize what's happening, I can make changes to both the pod configuration and our application so that I'm able to set `ResourceLimits` based on the pod's _actual_ memory availability. That leaves me with a couple of questions:

1. Is this a bug in the library? 
2. What is the recommended value for `ResourceLimits.Memory` in this constrained environment (350mi), where the largest images we work with are 30mb?
    - does `ResourceLimits.Memory` tune the same knobs as `ResourceLimits.LimitMemory`? I can't use the latter since passing a percentage won't work 
3. Should I be setting `ResourceLimits.MaxMemoryRequest` as well?
     - What's the different between `Memory` and `MaxMemoryRequest`. The documentation doesn't really explain it to me 
4. Are there any other settings worth tweaking manually?

In case it matters:
- we also have a scratch disk attached to the pods that we assign to the library via `MagickNET.SetTempDirectory` as we were expecting to spill to disk.
- we are using `Magick.NET-Q8-AnyCPU` version `14.0.0`

**Comment by dlemstra** [comment]
I saw that you opened an issue about this with some more information but let me first answer your questions here:

> 1. Is this a bug in the library?

The native library uses `sysconf(_SC_PHYS_PAGES)` and that is also done inside ImageMagick. It doesn't look at either the cgroup v1 or v2 file because this software was written long before containers where a thing. But `ResourceLimits.LimitMemory` is a convenience method that was only added to this library so I will need to check if I can improve that method to check those cgroup files.

> 2. What is the recommended value for ResourceLimits.Memory in this constrained environment (350mi), where the largest images we work with are 30mb?
    - does ResourceLimits.Memory tune the same knobs as ResourceLimits.LimitMemory? I can't use the latter since passing a percentage won't work

The `ResourceLimits.LimitMemory` method sets both `Memory` and `Area` as described in the documentation of that method:

> Set the maximum percentage of <see cref="Memory"/> that can be used for image data. 
This also changes the <see cref="Area"/> limit to four times the number of bytes.`

The size of an image doesn't tell you anything about how much memory it needs. An image of `1kb` could need `1gb` and a `30mb` image might even need less when a lot of the date is skipped metadata. So there is not really a good way to answer your question. In your case I 

> 3. Should I be setting ResourceLimits.MaxMemoryRequest as well?
    What's the different between Memory and MaxMemoryRequest. The documentation doesn't really explain it to me

Below is the document for both methods:

```C#
/// <summary>
/// Gets or sets the max memory request in bytes. ImageMagick maintains a separate memory pool for large
/// resource requests. If the limit is exceeded when allocating pixels, the allocation is instead memory-mapped
/// on disk.
/// </summary>
ulong MaxMemoryRequest { get; set; }

/// <summary>
/// Gets or sets the pixel cache limit in bytes. Once this memory limit is exceeded, all subsequent pixels cache
/// operations are to/from disk. The default value of this is 50% of the available memory on the machine in 64-bit mode.
/// When running in 32-bit mode this is 50% of the limit of the operating system.
/// </summary>
ulong Memory { get; set; }
```

What could I do to make it more clear the `Memory` is only for the pixels of an image and `MaxMemoryRequest` is also used when we need to read other image data.

> 4. Are there any other settings worth tweaking manually?

I do think that you should want to set the memory limit with `ResourceLimits.LimitMemory` but I would also advise you to set both ResourceLimits.Width` and `ResourceLimits.Height` because you probably don't want to read images with ridiculous dimensions.

**Reply by pinkfloydx33** [reply]
Thanks for the response and my apologies for opening the secondary issue.

Regarding the documentation, I've read all the xmldoc for the various properties. I guess I get confused between the difference between what controls the "pixel cache" (and what that even means) and what is an overall memory _limit_. `MaxMemoryRequest` for example seems to refer to both pixels and general memory. `LimitMemory` talks about _image data_, `Disk` refers to pixel cache bytes and "requests for memory" (noting that exceeding it will fail).

I've also read a few discussions on here where the terms were explained, but then left me even more confused (sorry I don't have links off hand). 

I guess there's just a lot a terminology here. I don't know if these are things I'd understand better if I was a user of the _actual_ Image Magick application, or if I'm just not reading carefully.

---
I've worked around the original issue by making some changes to my application. Kubernetes allows setting enviroment variables for things like the memory limit. My app will look for that and then compute the percentage itself. I then set `Memory` and `MaxMemoryRequest` to that value (based on a recc I saw elsewhere) and `Area` four-times that to match the behavior of LimitMemory.

I've also reduced the `MaxProfileSize` to 64kb. I will look at setting `Width` and `Height` per your recommendation.

**Reply by dlemstra** [reply]
Good that you also pointed `MaxProfileSize` out here. That setting is relatively new but I would also recommend that you set this. I think you could limit to `1mb` instead of `64kb`.

And good news I was able to make a change the `LimitMemory` implementation and it now checks use the cgroup memory limit when this is available. This will become available in the next release.

**Reply by pinkfloydx33** [reply]
@dlemstra thanks. We'll be pulling in that update and removing our workaround of calculating the memory ourselves.

I have a stupid question. Is LimitMemory for the entire NET process or is it per operation/invocation of image magick, or possibly per thread? I would imagine it's the former but I'm hitting issues with concurrent requests that's causing me to believe it's the latter somehow. I swear I've read through the docs but it's certainly possible I've missed something, so apologies if I have. 

When processing a single image everything is fine. But if we send multiple image uploads as seperate, concurrent HTTP requests we almost certainly end up getting OOM killed by Kubernetes. This happens even if I set LimitMemory (or equivalent) to something exceptionally low (it just takes more requests). I've taken gcdumps of the process right before it gets killed using dotnet-monitor and the managed memory is well within the expected bounds, which tells me that we are eating up native memory. I can do the same with many giant documents/non-images which all get processed the same way sans the thumbnail generation, so I've definitely ruled out everything else.

Is Magick.NET not able to handle this use case?  Or have I just not found the appropriate values for these limits? I'm hoping it's the latter; in the meantime I'm using a semaphore to ensure I've only got one operation with Image Magick at a time in a single pod which seems to resolve the issue.

**Reply by dlemstra** [reply]
The LimitMemory is for the entire .NET process and not per thread. You should get more information with the logging of ImageMagick:

```C#
MagickNET.SetLogEvents(LogEventTypes.Resource);
MagickNET.Log += (sender, args) =>
{
  Console.WriteLine(args.Message);
};
```

And are you using the `using` statement in all locations?

**Reply by pinkfloydx33** [reply]
Yes, everything is being properly disposed. But from what I can tell _thus far_ we aren't even getting that far (ie. two threads hit at the same time and we get OOM killed before they even finish processing).

I had originally set up logging but used the all/trace mode which was incredibly noisy. I'll try with just Resource. Is there anything specific I should look for?

Thanks again!

**Reply by dlemstra** [reply]
Did you also check what kind of input images you have? And by that I mean the dimensions of the image? You should see three numbers in the logging:  `[resource_request]/[resource_current]/[resource_limit]`.

**Reply by pinkfloydx33** [reply]
We're using 14.2.0 but if I pass 20% to `Limits.LimitMemory` and then inspect `Limits.Memory` it returns `2gb` which is 20% of `13gb` and back to the original issue I posted about.

```csharp
Limits.LimitMemory(new Percentage(20));
Console.WriteLine("{0}bytes", Limits.Memory);  // 2930716672bytes
```

**Reply by pinkfloydx33** [reply]
I've been able to slam my app concurrently with pretty much any Jpegs and _some_ PNGs/Gifs (of all sizes) without issue. I can seem to replicate it with _certain_ PNGs, HEIC and animated gifs, but what they have in common (if anything) I cannot figure out. I added in the logging like you suggested and while I don't fully understand them, I do see it acquiring and releasing. For the most part the request/current/limits look to be about what I'd expect.

I have noticed a few things that seem to be true when I get OOM Killed (for PNG/HEIC files anyways):

1. The disk caching (Magick Temp Directory) is being used
2. `Map` appears in the logs

I do get some log messages that look wrong to me. These `Map` logs still have a very large limit which equates to 2x the value of MemTotal (or my node's total memory)

```
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Map: 68.665MiB/68.665MiB/27.294GiB 
```

I also get these `Disk` logs which indicate `8EiB` as a maximum:

```
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Disk: 68.665MiB/68.665MiB/8EiB  
```

I think that's just `uint.MaxValue / 2` and is something I should set myself via `Limits.Disk`.  But I assume that would be unrelated to any memory issues.

**Reply by dlemstra** [reply]
The Map resource limit maps a memory request to a file on disk instead. And Disk is used when memory cannot be allocated and a file is created to contain the memory. They work similar but have a different api. What happens when you set `Map` to zero? Maybe that is also counted as memory when you run into an OOM?

**Reply by pinkfloydx33** [reply]
First, I wanted to thank you for your continued support here. This is really baffling us and I do appreciate you taking the time to respond. 

I couldn't find a way to set `Map` to _anything_; it was the first thing I looked for when I saw those logs. Am I just missing it?

---

If I'm on 14.2.0 shouldn't `Map` be picking up your fixes to account for cgroups? I assume it's sourced the same way, but the 27gb seems to indicates otherwise (though I suppose if LimitMemory isn't picking it up, Map wouldnt either)

**Reply by dlemstra** [reply]
It looks like I never added Map to the api, not sure why I didn't. Might be because there was no support for memory mapped files at an earlier point in time. Will need to investigate this and check why I did not add it.

**Comment by pinkfloydx33** [comment]
You mentioned to make sure I'm disposing correctly. I see the "release" in the logs, and I believe my code looks correct, but I'm wondering if we are perhaps doing something just plain wrong w/r/t this library. I would expect as many relinquish messages in the logs as acquire messages, but that's not what I see, so perhaps we are missing something. 

We generate thumbnails only. We resize the target image, decide on what format we want to convert it to and then export the bytes and try to compress. If the produced image is over some threshold and was a PNG with transparency, we convert it to a JPG instead. Here's the code we are running (error handling/our logging removed and everything moved into a single method).

<details>
<summary>csharp code</summary>

```csharp
internal static byte[] GetThumbnail(ReadOnlyMemory<byte> input)
{
    const uint maxSizeBytes = 65536;
    const uint maxDim = 256;

    // load image
    using var image = new MagickImage(input.Span);

    // resize to our target thumbnail size
    var size = new MagickGeometry(maxDim)
    {
        IgnoreAspectRatio = false,
        Greater = true,
        Less = false
    };
    image.Resize(size);

    // set target format and quality
    // GIF->GIF, Anything With Transparency->Png, Jpeg for everything else
    (image.Format, image.Quality) = image.Format switch
    {
        MagickFormat.Gif or MagickFormat.Gif87 => (MagickFormat.Gif, 100),
        _ when image.HasAlpha => (MagickFormat.Png, 97),
        _ => (MagickFormat.Jpeg, 70)
    };
    image.Strip();

    // get bytes compressing if possible
    // if the size is less than some max we define, then return it
    var bytes = GetBytesAndCompress(image);
    if (bytes.Length <= maxSizeBytes)
        return bytes;

    // if we've got a GIF/JPEG over the limits, return empty
    if (image.Format != MagickFormat.Png)
        return [ ];

    // now try and convert a PNG (with transparency) to a JPEG
    // setting the background color and removing the Alpha
    image.BackgroundColor = MagickColors.White;
    image.Alpha(AlphaOption.Remove);
    image.Format = MagickFormat.Jpg;
    image.Quality = 70;
    image.Strip();

    // get bytes and once again try and compress
    bytes = GetBytesAndCompress(image);

    // return bytes if less than max, otherwise empty
    return bytes.Length <= maxSizeBytes ? bytes : [ ];
}

static readonly ImageOptimizer Optimizer = new() { IgnoreUnsupportedFormats = true, };
static byte[] GetBytesAndCompress(MagickImage image)
{
    var bytes = image.ToByteArray(image.Format);

    using var ms = new MemoryStream(bytes);
    if (Optimizer.Compress(ms))
        return ms.ToArray();

    return bytes;
}
```
</details>

Now when I run a Jpeg through these are the logs I get (slightly redacted and the timestamps removed to reduce clutter):

<details>
<summary>logs</summary>

```
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource 
  Area: 12.759MiB/0B/16.25MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Memory: 38.276MiB/38.276MiB/162.5MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Area: 12.759MiB/0B/16.25MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Memory: 51.035MiB/89.311MiB/162.5MiB
Resource My.App[1]: resource.c/RelinquishMagickResource/1126/Resource
  Memory: 38.276MiB/51.035MiB/162.5MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Area: 25344B/0B/16.25MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Memory: 76032B/51.107MiB/162.5MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Area: 569.54KiB/0B/16.25MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Memory: 1.6686MiB/52.776MiB/162.5MiB
Resource My.App[1]: resource.c/RelinquishMagickResource/1126/Resource
  Memory: 1.6686MiB/51.107MiB/162.5MiB
Resource My.App[1]: resource.c/RelinquishMagickResource/1126/Resource
  Memory: 51.035MiB/76032B/162.5MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Memory: 76032B/148.5KiB/162.5MiB
Resource My.App[1]: resource.c/RelinquishMagickResource/1126/Resource
  Memory: 76032B/76032B/162.5MiB
Resource My.App[1]: resource.c/RelinquishMagickResource/1126/Resource
  Memory: 76032B/0B/162.5MiB
```
</details>

Notice that there's 5 "relinquish" messages to the 9 "acquire" messages. I'm not sure if that's expected or not. 

Does the above show us doing anything horribly wrong that would cause a native memory leak?

**Reply by dlemstra** [reply]
As you might have noticed the Area acquire does not increment the "current" value it's a limit per request. In the log entry `12.759MiB/0B/16.25MiB` the second value remains zero. And in the last Relinquish message you will notice that the current memory value becomes zero again so there is no leak.

**Reply by pinkfloydx33** [reply]
First, I really want to thank you again for your assistance and putting up with me on this.

I realize it's not very scientific, but I connected to my pod and output the contents of `/sys/fs/cgroup/memory.current` in a loop. If I skip image processing and just send the file to blob storage the current memory usage typically only goes up by the file size and then gets recovered by GC and stays fairly constant. When I actually process the images I have cases where I can see the used memory go up but never comes back all the way down. 

It's all dependent on the image of course. JPEGs seem well behaved. The memory use goes up and then comes down and I can send a big batch of them (50 concurrently) without worrying about the process being killed. 

Take this JPEG for example. 
<details>
<summary>6000x4000 / 7,805KB / JPG</summary>

![1693761761](https://github.com/user-attachments/assets/31bf2c0c-2d2e-4b16-8e7a-c0dc2dd44e3b)

</summary>
</details>

It produces this memory usage output:

```
412078080 <-- before
415444992 <-- request
461926400
500064256
418054144 <-- done
```

It does make a jump there consistent with the file size, but as you can see it recovers. Even if I send 50 concurrent requests of that same image I don't get OOM Killed. The process comes close though, at one point jumping up to 648mb but it seems to cap out and throttle the requests before relinquishing the memory and eventually getting GC'd, ending up back where I started.

Now take this HEIC file which seems to be my real problem child. (GitHub wouldn't let me upload a .heic so I had to .zip it)
[IMG_4976.zip](https://github.com/user-attachments/files/18067161/IMG_4976.zip)

The memory usage makes an increase equivalent to the file's size, but then never drops back down
```
486514688 <-- before
486629376 <-- request
532353024 
527298560 <-- done
527339520 
527208448
```

It then stays there **forever**. If I send the file 3-4 more times (or try to send 4 concurrently) the memory use climbs considerably, **never** recovers and I get OOM killed. But that said, the logs look fine:

<details>
<summary>logs</summary>

```
Resource My.App[1]: resource.c/AcquireUniqueFileResource/578/Resource
  ...
Resource My.App[1]: resource.c/AcquireUniqueFileResource/635/Resource
  /app/image-scratch/magick-XuA_Vf3uOlqLzZQL7KTvZj3GnBkuoMq2
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  ListLength: 1/0/1000
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Area: 11.628MiB/0B/16.25MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Memory: 34.884MiB/34.884MiB/65MiB
Resource My.App[1]: resource.c/RelinquishUniqueFileResource/1165/Resource
  /app/image-scratch/magick-XuA_Vf3uOlqLzZQL7KTvZj3GnBkuoMq2
Resource My.App[1]: resource.c/RelinquishMagickResource/1126/Resource
  Memory: 34.884MiB/0B/65MiB
Resource My.App[1]: resource.c/AcquireUniqueFileResource/578/Resource
  ...
Resource My.App[1]: resource.c/AcquireUniqueFileResource/635/Resource
  /app/image-scratch/magick-sk6_4YwyjCU8QHOCpguRTw3Y8Fy9Fs4Y
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  ListLength: 1/0/1000
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Area: 11.628MiB/0B/16.25MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Memory: 34.884MiB/34.884MiB/65MiB
Resource My.App[1]: resource.c/RelinquishUniqueFileResource/1165/Resource
  /app/image-scratch/magick-sk6_4YwyjCU8QHOCpguRTw3Y8Fy9Fs4Y
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  ListLength: 1/0/1000
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Area: 49152B/0B/16.25MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Memory: 144KiB/35.024MiB/65MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  ListLength: 1/0/1000
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Area: 756KiB/0B/16.25MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Memory: 2.2148MiB/37.239MiB/65MiB
Resource My.App[1]: resource.c/RelinquishMagickResource/1126/Resource
  Memory: 2.2148MiB/35.024MiB/65MiB
Resource My.App[1]: resource.c/RelinquishMagickResource/1126/Resource
  Memory: 34.884MiB/144KiB/65MiB
Resource My.App[1]: resource.c/AcquireMagickResource/384/Resource
  Memory: 144KiB/288KiB/65MiB
Resource My.App[1]: resource.c/RelinquishMagickResource/1126/Resource
  Memory: 144KiB/144KiB/65MiB
Resource My.App[1]: resource.c/RelinquishMagickResource/1126/Resource
  Memory: 144KiB/0B/65MiB
```
</details>

I even tried setting `Limits.Memory` and `Limts.MaxMemoryRequest` to `0` which is what I think you might have been suggesting above it still exhibits the same behavior--it just takes a few extra files to happen. I know it's not a very scientific test but doesn't that jump in usage that doesn't recover show _something_ being amiss?

I also get real weird behavior with a specific GIF file:

<details>
<summary> 300x300 / 110KB / Animated GIF (flashes!) </summary>

![totoro-transparent](https://github.com/user-attachments/assets/d1be9d6d-9656-4cfe-8d17-5b4dcd96d98a)

</details>

If I set `Limits.Memory` to the _equivalent_ of 10-15% of my pod's memory then I can upload and process this file all day long, even 50+ concurrently and the memory usage just caps out and recovers. If I set `Limits.Memory` to the equivalent of 20% instead, then sending this file will eventually OOM kill the process. Even sending just `4` concurrently will do it. If I monitor the memory usage directly it behaves similar to the HEIC. Even weirder is that _sometimes_ if it's the absolute first request, it will blow up instantly. I have yet to be able to capture that case with logs though. 

---

It seems so far that that I am better off keeping the memory limit low, around 10% which I guess just means most things will end up getting sent to disk. It has solved most of the issues I am having, for example that reduction has allowed me to send 30 concurrent uploads of a 20mb PNG file when otherwise I couldn't send more than 2. The only thing it's yet to help is the issue with the HEIC.

For whatever it's worth, here's how I'm setting the current Limits:

- Actual Pod Memory Limit: 681,574,400
- Memory: 68,157,440 (manually calculated 10%)
- MaxMemoryRequest: 68,157,440
- Area: 17,039,360
- ListLength: 1,000
- Height: 30,000
- Width: 30,000
- MaxProfileSize: 1Mb
- Disk: 1,048,576,000  (1Gb)

Unfortunately I have no control over what my users upload. I need to support "common" formats, but also iPhone formats (hence the HEIC). Other than setting limits, is there anything I can do to optimize these flows or even to simply stop processing in its tracks if the file has a good chance of causing issues?

**Reply by dlemstra** [reply]
It looks like you are running out of memory because of managed code instead. You are allocating multiple arrays and .NET will not directly return memory to the operating system. And then you also have unmanaged memory being allocated. Combining that with the managed memory could result in out of memory issues. Maybe you should also introduce some limits on the managed side yourself?

**Reply by pinkfloydx33** [reply]
I'm not sure what else I can do on the managed side other than limiting files sizes like we already are, but you've already pointed out that that doesn't really matter. We're already using Span/Memory and avoiding allocations as much as we can. 

Setting the memory limit low seems to solve most of my problems since I guess that just means it's doing more work via the disk. That's absolutely fine for our use case. We aren't dealing with high load here so if it takes a performance hit that's no big deal to us. The problem really is that _over time_ something is happening that eventually pushes us over that native + managed limit. Kubernetes is indiscriminate in it's OOM kill. If there were some way to gracefully terminate in this case then I'd just deal with it, but there's not and we end up with some form of data corruption. 

I've gotten it to the point where I can process 50+ 20mb PNG files _concurrently_ without any issue (the memory spikes, but the GC collects)... yet I can't process five 2mb HEIC files **over the lifetime** of my application no matter how I tune the limits. Just one request causes the memory usage of the app to increase by ~40mb _forever_, whereas without image processing it only increases _temporarily_ by 2mb. I realize that the logs look fine and I'm not going to convince you otherwise, but something just feels wrong about that.  And unless there's something egregious in my code sample above, there's really nothing left that I _can_ do.

> It looks like you are running out of memory because of managed code instead

That's what I had originally thought might be the issue. GC heap dumps and runtime metrics aren't showing that that's the case though. The GC is definitely collecting as it gets pressure. Even with tons of requests the usage spikes (approximately by the file sizes) but then gets collected in short order if I'm sending lots of requests. The arrays don't show up in the memory dump outside of what the array pool caches per TLS and if I take Magick.NET out of the picture everything is fine no matter how hard I hammer. 

I've forced GC, including the LOH and it helps a little. Setting the GC heap hard limit to something lowish forces collections more frequently which helps a bit as well. I've also tried Workstation GC since that doesn't pre-allocate as much, but that only prolongs the inevitable. 

---

I apologize for the spam, and once again thank you for your time. I've been trying to resolve this for a few weeks now.  It's a huge PITA to try and troubleshoot since I can't do it on my local machine. I am grasping at straws and growing increasingly frustrated. If I upload many more files or review just one more memory dump I might lose it soon 😛

**Reply by dlemstra** [reply]
This weekend I ran your heic image through a memory analyzer and no leaks were reported. I also tried to run your code to see memory increments and I did not notice anything strange. The only thing that I noticed with local debugging was that I had to disable OpenCL because OpenCL acceleration would require extra host memory. But this is disabled by default and you did not mention that you turned it on.

**Reply by pinkfloydx33** [reply]
Yeah, I don't even know what that is 😄

I'm sure you're right that it's something else in our code, but I've been unable to figure out what. I was wondering if it could be Linux distro specific or something but that's the only thought I have left.

That said, we've decided to go a different route for now given that I've spent so much time on this. I've added a liveness probe to our app that will cause Kubernetes to gracefully restart the container if it's using more than X percent of overall memory for Y duration. At least then the app recovers and we have enough replicas for that to hopefully not be an issue. In my load testing it's performed quite well. 

It's definitely a hammer, but I've spent enough time trying to figure it out unfortunately. We've also switched to DATAS (default in net9) which gives workstation-GC-like memory consumption, and we've enabled conservativeGC and lowered the GC high-memory percent. All of which at least prolong the overall issue a good deal as well.

Thanks again for your assistance. I really do appreciate it.
